modules/http/src/main/java/org/glassfish/grizzly/http/util/HttpDateFormat.java

```diff
@@ -0,0 +1,195 @@
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static org.glassfish.grizzly.http.util.HttpCodecUtils.toCheckedByteArray;
+
+/**
+ * Utility class to generate HTTP dates.
+ * This class is based on {@link FastHttpDateFormat},
+ * but uses java.time API instead of the old java.util.Date/Calendar and java.text.DateFormat/SimpleDateFormat API.
+ *
+ * @author Bongjae Chang
+ */
+public final class HttpDateFormat {
+
+    private static final String ASCII_CHARSET_NAME = Charsets.ASCII_CHARSET.name();
+
+    private static final int CACHE_SIZE = 1000;
+
+    private static final ZoneId GMT_ZONE_ID = ZoneId.of("GMT");
+
+    // Fri, 03 Oct 2025 04:47:00 GMT
+    private static final DateTimeFormatter FORMATTER =
+            DateTimeFormatter.ofPattern("EEE, dd MMM yyyy HH:mm:ss zzz", Locale.US).withZone(GMT_ZONE_ID);
+
+    private static final DateTimeFormatter[] FORMATTERS = new DateTimeFormatter[]{
+            DateTimeFormatter.ofPattern("EEE, dd MMM yyyy HH:mm:ss zzz", Locale.US).withZone(GMT_ZONE_ID),
+            DateTimeFormatter.ofPattern("EEE, dd-MMM-yy HH:mm:ss zzz", Locale.US).withZone(GMT_ZONE_ID),
+            DateTimeFormatter.ofPattern("EEE MMMM d HH:mm:ss yyyy", Locale.US).withZone(GMT_ZONE_ID)};
+
+    /**
+     * Instant on which the currentDate object was generated.
+     */
+    private static volatile long nextGeneration;
+
+    private static final AtomicBoolean isGeneratingNow = new AtomicBoolean();
+
+    /**
+     * Current formatted date as byte[].
+     */
+    private static volatile byte[] currentDateBytes;
+
+    /**
+     * Current formatted date.
+     */
+    private static volatile String cachedStringDate;
+    private static volatile byte[] dateBytesForCachedStringDate;
+
+    /**
+     * Formatter cache.
+     */
+    private static final ConcurrentMap<Long, String> formatCache = new ConcurrentHashMap<>(CACHE_SIZE, 0.75f, 64);
+
+    /**
+     * Parser cache.
+     */
+    private static final ConcurrentMap<String, Long> parseCache = new ConcurrentHashMap<>(CACHE_SIZE, 0.75f, 64);
+
+    // --------------------------------------------------------- Public Methods
+
+    /**
+     * Get the current date in HTTP format.
+     */
+    public static String getCurrentDate() {
+        final byte[] currentDateBytesNow = getCurrentDateBytes();
+        if (currentDateBytesNow != dateBytesForCachedStringDate) {
+            try {
+                cachedStringDate = new String(currentDateBytesNow, ASCII_CHARSET_NAME);
+                dateBytesForCachedStringDate = currentDateBytesNow;
+            } catch (UnsupportedEncodingException ignored) {
+                // should never reach this line
+            }
+        }
+        return cachedStringDate;
+    }
+
+    /**
+     * Get the current date in HTTP format.
+     */
+    public static byte[] getCurrentDateBytes() {
```

Comment:
IMO this method is dangerous.

Please try something similar (if I haven't got anything wrong)

```java
public static byte[] getCurrentDateBytes() {
        final long now = System.currentTimeMillis();
        final long diff = now - nextGeneration;

        if (diff < 0 || (diff < 5000 && isGeneratingNow.get())) {
            return currentDateBytes;
        }

        while (!isGeneratingNow.compareAndSet(false, true)) {
            Thread.onSpinWait();
        }

        if (now > nextGeneration) {
            currentDateBytes = toCheckedByteArray(new StringBuilder(FORMATTER.format(Instant.ofEpochMilli(now))));
            nextGeneration = now + 1000;
        }

        isGeneratingNow.set(false);

        return currentDateBytes;
    }
```

Reply:
Your suggested approach is probably correct. I'll look into it a bit more. I know it's risky, but I allowed it because the logic didn't seem to prioritize consistency. However, I think my approach wasn't a good one.